pkg/controller/admissionchecks/provisioning/controller.go

```diff
@@ -0,0 +1,715 @@
+	"context"
+	"crypto/sha1"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"maps"
+
+	corev1 "k8s.io/api/core/v1"
+	apimeta "k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/sets"
+	autoscaling "k8s.io/autoscaler/cluster-autoscaler/provisioningrequest/apis/autoscaling.x-k8s.io/v1beta1"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/utils/ptr"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	kueue "sigs.k8s.io/kueue/apis/kueue/v1beta1"
+	utilmaps "sigs.k8s.io/kueue/pkg/util/maps"
+	"sigs.k8s.io/kueue/pkg/util/slices"
+	"sigs.k8s.io/kueue/pkg/workload"
+)
+
+const (
+	objNameHashLength = 5
+	// 253 is the maximal length for a CRD name. We need to subtract one for '-', and the hash length.
+	objNameMaxPrefixLength = 252 - objNameHashLength
+	podTemplatesPrefix     = "ppt"
+)
+
+var (
+	errInconsistentPodSetAssignments = errors.New("inconsistent podSet assignments")
+)
+
+type Controller struct {
+	client client.Client
+	cache  *cache
+}
+
+var _ reconcile.Reconciler = (*Controller)(nil)
+
+// +kubebuilder:rbac:groups="",resources=events,verbs=create;watch;update
+// +kubebuilder:rbac:groups="",resources=podtemplates,verbs=get;list;watch;create;delete;update
+// +kubebuilder:rbac:groups=autoscaling.x-k8s.io,resources=provisioningrequests,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=autoscaling.x-k8s.io,resources=provisioningrequests/status,verbs=get
+// +kubebuilder:rbac:groups=kueue.x-k8s.io,resources=workloads,verbs=get;list;watch;update;patch;delete
+// +kubebuilder:rbac:groups=kueue.x-k8s.io,resources=workloads/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=kueue.x-k8s.io,resources=admissionchecks,verbs=get;list;watch
+// +kubebuilder:rbac:groups=kueue.x-k8s.io,resources=provisioningrequestconfigs,verbs=get;list;watch
+
+func NewController(client client.Client) *Controller {
+	return &Controller{
+		client: client,
+		cache: &cache{
+			checks:  make(map[string]*kueue.AdmissionCheck),
+			configs: make(map[string]*kueue.ProvisioningRequestConfig),
+		},
+	}
+}
+
+// Reconcile performs a full reconciliation for the object referred to by the Request.
+// The Controller will requeue the Request to be processed again if an error is non-nil or
+// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
+func (c *Controller) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	wl := &kueue.Workload{}
+	log := ctrl.LoggerFrom(ctx)
+	log.V(2).Info("Reconcile workload")
+
+	err := c.client.Get(ctx, req.NamespacedName, wl)
+	if err != nil {
+		return reconcile.Result{}, client.IgnoreNotFound(err)
+	}
+
+	if !workload.HasQuotaReservation(wl) {
```

Comment:
Does it also handle workloads that are finished?

Reply:
this was changed